add persistent volume to crowdsec chart

flo-mic opened the pull request:
This adds a persistent volume to the crowdsec helm chart so the database can be persistent and will survive a pod update. Currently if you destroy the lapi pod, the list of known bouncers is destroyed and you need to register them again.

he2ss commented:
Hi @flo-mic,

Thanks for your contributions ! It will help a lot.
For this PR, it seems the user need to specify the `storageClass` name (specific to user context) in it's crowdsec-values.yaml file.
Isn't better to disabled it by default, and if the user wants to enable, he need to specify the `storageClass` name, size etc ?

flo-mic commented:
@he2ss the user does not need to specify the `storageClass`. Currently the storageClass is "", which means that kubernetes will use the default storage provider (if some storage is flagged as default with `metadata.annotations.storageclass.kubernetes.io/is-default-class: "true"`). With the curreent implementation the user also hast he possibility to choose another storage class if he does not want to use the default one. In this case the value "" needs to be replaced with the storage class to use.

Regarding the size, I used 1Gi for the beginning as from my experience the storage requirements are not that high. Nevertheless, most storage provider support volume expansion and will manage an expansion automatically.

he2ss commented:
@flo-mic, I just tested your changes, it's working quite well. There is still some issues we need to improve if you want to add them :

* Add the persistent volume By default without having to enable the dashboard.
* Add persistent volume for `/etc/crowdsec/` or at least `/etc/crowdsec/online_api_credentials.yaml` for the online registration.

If you want we can merge this PR and I'll do directly after those changes to improve the persistency.

flo-mic commented:
@he2ss so I just fixed the dashboard part. Sorry did not see this before as my dashboard was always enabled. Also a second PVC for config files is now added and I changed the version number for the release.

he2ss commented on charts/crowdsec/templates/lapi-deployment.yaml:
```diff
@@ -72,12 +80,25 @@ spec:
+      {{- end }}
       terminationGracePeriodSeconds: 30
-      {{ if .Values.lapi.dashboard.enabled }}
+      {{- if or (.Values.lapi.persistentVolume.data.enabled) (.Values.lapi.persistentVolume.config.enabled) (.Values.lapi.dashboard.enabled) }}
```
this check is not needed here `(.Values.lapi.persistentVolume.config.enabled)` as the crowdsec config volume isn't used in this container.

flo-mic replied:
the crowdsec config file is used in the container crowdsec-lapi or? then we need to load it in the pod or? (We only load it in the crowdsec-lapi container and not in the dashboard, but the volume must be available on pod level)

he2ss replied:
Yes the crowdsec config is only used on the crowdsec-lapi container. You already add it as volume in L66.
VolumesMounts are per container and not pod no ?

he2ss replied:
Yes the crowdsec config is only used on the crowdsec-lapi container. You already add it as volume in L66.
VolumesMounts are per container and not pod no ?

flo-mic replied:
I think we have a misunderstanding here. Maybe you have misread the line intends? The current structure looks something like this (deleted a lot of content for simplicity):
```
apiVersion: apps/v1
kind: Deployment
spec:
  template:
    spec:
      initContainers:
        - name: fetch-metabase-config
      containers:
        - name: crowdsec-lapi        
          volumeMounts:
            - name: crowdsec-db
              mountPath: /var/lib/crowdsec/data
            - name: crowdsec-config
              mountPath: /etc/crowdsec
        - name: dashboard
          volumeMounts:
          - name: shared-data
            mountPath: /metabase-data
          - name: crowdsec-db
            mountPath: /var/lib/crowdsec/data
      volumes:
      - name: crowdsec-db
        persistentVolumeClaim:
          claimName: {{ .Release.Name }}-db-pvc
      - name: crowdsec-config
        persistentVolumeClaim:
          claimName: {{ .Release.Name }}-config-pvc
```

We define the pvc in `spec.template.spec.volumes` as a `volume` and give it a name. At this moment the pvc is defined but it is not visible to any container. Then we have two containers here, the metabase dashboard and the lapi container, but only the lapi container can access the volume as the `spec.template.spec.containers.[crowdsec-lapi].volumeMounts` contains a `volumeMount` for the defined volume `crowdsec-config`. The dashboard does not have it and so it can not access the storage.

he2ss replied:
Ah sorry for the misunderstanding. You are actually right :+1:.

he2ss commented on charts/crowdsec/templates/lapi-deployment.yaml:
```diff
@@ -72,12 +80,25 @@ spec:
-      {{ if .Values.lapi.dashboard.enabled }}
+      {{- if or (.Values.lapi.persistentVolume.data.enabled) (.Values.lapi.persistentVolume.config.enabled) (.Values.lapi.dashboard.enabled) }}
       volumes:
+      {{- if .Values.lapi.dashboard.enabled }}
```
This check isn't needed, as we already check the same variable to deploy all the container resource.

flo-mic replied:
Sorry I don't get this, for me it looks right. We are not inside a container in that line. We need to define the shared-data volume in case the dashboard is enabled. But on this level we can not guarantee that the dashboard is enabled

he2ss replied:
Sorry I was not clear about it. The thing is that check `{{- if .Values.lapi.dashboard.enabled }}` is already in line 71 https://github.com/crowdsecurity/helm-charts/blob/b24bd8dd1573e9496073b54e0e97ee3d9a74b0ec/charts/crowdsec/templates/lapi-deployment.yaml#L71 

So yes it's still correct but just redundant.

he2ss replied:
Sorry I was not clear about it. The thing is that check `{{- if .Values.lapi.dashboard.enabled }}` is already in line 71 https://github.com/crowdsecurity/helm-charts/blob/b24bd8dd1573e9496073b54e0e97ee3d9a74b0ec/charts/crowdsec/templates/lapi-deployment.yaml#L71 

So yes it's still correct but just redundant.

flo-mic replied:
I think this is also resolved or? We need the check to decide if we want to deploy the dashboard or not. There is no other check before beside of the one for the init container

he2ss replied:
Sorry for the delay. I just tested the helm chart too. Seems good :+1:

he2ss commented on charts/crowdsec/templates/lapi-deployment.yaml:
```diff
@@ -72,12 +80,25 @@ spec:
-      {{ end }}
\ No newline at end of file
+      {{- end }}
+      {{- if .Values.lapi.persistentVolume.config.enabled }}
```
This volume isn't used here, we can just avoid mounting it in this container (the metabase dashboard only needs crowdsec database)

he2ss reviewed (commented):
Just small changes left. Otherwise the PR is so helpful thanks :slightly_smiling_face:.

flo-mic commented:
I fixed two additional issues which I just notied during my testing.
1. on an helm upgrade the storage class is not added to the PVC object if it is `""` and the deployment failed. This is a bug in helm 3, see https://github.com/helm/charts/issues/20389#issuecomment-583854511 
2. The mounted volume in `etc/crowdsec` was empty and the container could not start. Therefore the fix mounts the volume in another location, moves all missing files to the mount (if not present) and creates a link between the folders.

he2ss reviewed (approved):
Well done and thanks :)
